crates/dbsp/src/trace/mod.rs

```diff
@@ -444,6 +445,51 @@ where
         batcher.seal()
     }
 
+    /// Creates a new batch as a copy of `batch`, using `timestamp` for all of
+    /// the new batch's timestamps This is useful for adding a timestamp to a
+    /// batch, or for converting between different batch implementations
+    /// (e.g. writing an in-memory batch to disk).
+    ///
+    /// TODO: for adding a timestamp to a batch, this could be implemented more
+    /// efficiently by having a special batch type where all updates have the same
+    /// timestamp, as this is the only kind of batch that we ever create directly in
+    /// DBSP; batches with multiple timestamps are only created as a result of
+    /// merging.  The main complication is that we will need to extend the trace
+    /// implementation to work with batches of multiple types.  This shouldn't be
+    /// too hard and is on the todo list.
+    fn from_batch<BI>(batch: &BI, timestamp: &Self::Time, factories: &Self::Factories) -> Self
+    where
+        BI: BatchReader<Key = Self::Key, Val = Self::Val, Time = (), R = Self::R>,
+    {
+        Self::from_cursor(batch.cursor(), timestamp, factories, batch.len())
+    }
+
+    /// Creates a new batch as a copy of the tuples accessible via `cursor``,
+    /// using `timestamp` for all of the new batch's timestamps.
+    fn from_cursor<C>(
+        mut cursor: C,
+        timestamp: &Self::Time,
+        factories: &Self::Factories,
+        capacity: usize,
+    ) -> Self
+    where
+        C: Cursor<Self::Key, Self::Val, (), Self::R>,
+    {
+        let mut builder = Self::Builder::with_capacity(factories, timestamp.clone(), capacity);
+        let mut weight = cursor.weight_factory().default_box();
+        while cursor.key_valid() {
+            while cursor.val_valid() {
+                cursor.weight().clone_to(&mut weight);
+                if !weight.is_zero() {
```

Comment:
The deleted implementation was meant for individual batches, so this check wasn't necessary. We might have a slightly optimized implementation for that case that skips the check. I don't know if that's worth the trouble.

Reply:
I don't think a comparison against zero is going to have a significant cost, even in this inner loop, but yeah I know what you mean.

Reply:
it goes through at least one pointer indirection due to dynamic dispatch. But let's not optimize prematurely.

Reply:
I forgot about the indirection. I'm so used to Rust doing monomorphization.